snuba/query/parser/__init__.py

```diff
@@ -103,3 +128,75 @@ def _parse_query_impl(body: MutableMapping[str, Any], dataset: Dataset) -> Query
         having=having_expr,
         order_by=orderby_exprs,
     )
+
+
+def _validate_aliases(query: Query) -> None:
+    """
+    Ensures that no alias has been defined multiple times for different
+    expressions in the query. Thus rejecting queries with shadowing.
+    """
+    all_declared_aliases: Mapping[str, Set[Expression]] = defaultdict(set)
+    for exp in query.get_all_expressions():
+        # TODO: Make it impossible to assign empty string as an alias.
+        if exp.alias:
+            all_declared_aliases[exp.alias].add(exp)
+            new_exps = all_declared_aliases[exp.alias]
+            if len(new_exps) > 1:
+                raise ValueError(
+                    f"Shadowing aliases detected for alias: {exp.alias}. Expressions: {new_exps}"
+                )
+
+
+# A column name like "tags[url]"
+NESTED_COL_EXPR_RE = re.compile(r"^([a-zA-Z0-9_\.]+)\[([a-zA-Z0-9_\.:-]+)\]$")
+
+
+def _parse_subscriptables(query: Query) -> None:
+    """
+    Turns columns formatted as tags[asd] into SubscriptableReference.
+    """
+    current_aliases = {exp.alias for exp in query.get_all_expressions() if exp.alias}
+
+    def transform(exp: Expression) -> Expression:
+        if not isinstance(exp, Column):
+            return exp
+        match = NESTED_COL_EXPR_RE.match(exp.column_name)
+        if match is None or exp.column_name in current_aliases:
```

Comment:
Why not hoist the `column_name in current_aliases` bit up and avoid the `re.match` cost? It might have made it easier for my noob eyes to see, "ignore non-columns, ignore aliases, ignore non-subscripts" vs the or clause you have here.

If there's a reason not to, that's fine as well.